File: 3-HARD/Python/7-minimum-cost-max-flow/proposed-solution/min_cost_max_flow.py
from collections import deque
from typing import List, Tuple

class MinCostMaxFlow:
    def __init__(self, n: int):
        self.n = n
        self.graph = [[] for _ in range(n)]

    def add_edge(self, u: int, v: int, capacity: int, cost: int):
        self.graph[u].append([v, capacity, cost, len(self.graph[v])])
        self.graph[v].append([u, 0, -cost, len(self.graph[u]) - 1])

    def bellman_ford(self, source: int, dist: List[int], parent: List[Tuple[int, int]]):
        dist[:] = [float('inf')] * self.n
        dist[source] = 0
        in_queue = [False] * self.n
        queue = deque([source])
        in_queue[source] = True

        while queue:
            u = queue.popleft()
            in_queue[u] = False
            for i, (v, capacity, cost, rev) in enumerate(self.graph[u]):
                if capacity > 0 and dist[u] + cost < dist[v]:
                    dist[v] = dist[u] + cost
                    parent[v] = (u, i)
                    if not in_queue[v]:
                        queue.append(v)
                        in_queue[v] = True

    def min_cost_max_flow(self, source: int, sink: int) -> Tuple[int, int]:
        flow = 0
        cost = 0
        dist = [0] * self.n
        parent = [(-1, -1)] * self.n

        while True:
            self.bellman_ford(source, dist, parent)
            if dist[sink] == float('inf'):
                break

            path_flow = float('inf')
            v = sink
            while v != source:
                u, i = parent[v]
                path_flow = min(path_flow, self.graph[u][i][1])
                v = u

            flow += path_flow
            cost += path_flow * dist[sink]

            v = sink
            while v != source:
                u, i = parent[v]
                self.graph[u][i][1] -= path_flow
                rev = self.graph[u][i][3]
                self.graph[v][rev][1] += path_flow
                v = u

        return flow, cost

File: 3-HARD/JavaScript/7-minimum-cost-max-flow/proposed-solution/MinCostMaxFlow.js
// MinCostMaxFlow.js - Minimum Cost Maximum Flow algorithm using Bellman-Ford in JavaScript

class MinCostMaxFlow {
  constructor(n) {
    this.n = n;
    this.graph = Array.from({ length: n }, () => []);
  }

  addEdge(u, v, capacity, cost) {
    this.graph[u].push([v, capacity, cost, this.graph[v].length]);
    this.graph[v].push([u, 0, -cost, this.graph[u].length - 1]);
  }

  bellmanFord(source, dist, parent) {
    dist.fill(Infinity);
    dist[source] = 0;
    const inQueue = new Array(this.n).fill(false);
    const queue = [source];
    inQueue[source] = true;

    while (queue.length > 0) {
      const u = queue.shift();
      inQueue[u] = false;
      for (let i = 0; i < this.graph[u].length; i++) {
        const [v, capacity, cost] = this.graph[u][i];
        if (capacity > 0 && dist[u] + cost < dist[v]) {
          dist[v] = dist[u] + cost;
          parent[v] = [u, i];
          if (!inQueue[v]) {
            queue.push(v);
            inQueue[v] = true;
          }
        }
      }
    }
  }

  minCostMaxFlow(source, sink) {
    let flow = 0;
    let cost = 0;
    const dist = new Array(this.n);
    const parent = new Array(this.n).fill([-1, -1]);

    while (true) {
      this.bellmanFord(source, dist, parent);
      if (dist[sink] === Infinity) {
        break;
      }

      let pathFlow = Infinity;
      let v = sink;
      while (v !== source) {
        const [u, i] = parent[v];
        pathFlow = Math.min(pathFlow, this.graph[u][i][1]);
        v = u;
      }

      flow += pathFlow;
      cost += pathFlow * dist[sink];

      v = sink;
      while (v !== source) {
        const [u, i] = parent[v];
        this.graph[u][i][1] -= pathFlow;
        const rev = this.graph[u][i][3];
        this.graph[v][rev][1] += pathFlow;
        v = u;
      }
    }

    return { flow, cost };
  }
}

export default MinCostMaxFlow;
